src/flow/max-flow-min-cut.jl

```diff
@@ -0,0 +1,134 @@
+
+Use a default capacity of 1 when the capacity matrix isn't specified
+"""
+
+function maximum_flow{T<:Number}(
+    flow_graph::LightGraphs.DiGraph,       # the input graph
+    source::Int,                           # the source vertex
+    target::Int,                           # the target vertex
+    capacity_matrix::AbstractArray{T,2}=   # edge flow capacities
+        DefaultDistance()
+    )
+    n = size(flow_graph)[1]                # number of vertexes
+    flow = 0
+    flow_matrix = zeros(T, n, n)           # initialize flow matrix
+
+    while true
+        v, P, S = fetch_path(flow_graph, source, target, flow_matrix, capacity_matrix)
+
+        if P == None                       # no more valid paths
+            break
+        else
+            path = [v]                     # initialize path
+
+            u = v
+            while u!=source                # trace path from v to source
+                u = P[u]
+                push!(path, u)
+            end
+            reverse!(path)
+
+            u = v                          # trace path from v to target
+            while u!=target
+                u = S[u]
+                push!(path, u)
+            end
+                                           # augment flow along path
+            flow += augment_path!(path, flow_matrix, capacity_matrix)
+        end
+    end
+
+    return flow, flow_matrix
+end
+
+"""Calculates the amount by which flow can be augmented in the given path.
+Augments the flow and returns the augment value."""
+
+function augment_path!{T<:Number}(
+    path::Vector{Int},                     # input path
+    flow_matrix::AbstractArray{T,2},       # the current flow matrix
+    capacity_matrix::AbstractArray{T,2}    # edge flow capacities
+    )
+    augment = typemax(T)                   # initialize augment
+    for i in 1:length(path)-1              # calculate min capacity along path
+        u = path[i]
+        v = path[i+1]
+        augment = min(augment,capacity_matrix[u,v] - flow_matrix[u,v])
+    end
+
+    for i in 1:length(path)-1              # augment flow along path
+        u = path[i]
+        v = path[i+1]
+        flow_matrix[u,v] += augment
+        flow_matrix[v,u] -= augment
+    end
+
+    return augment
+end
+
+"""Uses Bidirectional BFS to look for augmentable-paths. Returns the vertex where
+the two BFS searches intersect, the Parent table of the path as well as the
+Successor table of the path found."""
+
+function fetch_path{T<:Number}(
+    flow_graph::LightGraphs.DiGraph,       # the input graph
+    source::Int,                           # the source vertex
+    target::Int,                           # the target vertex
+    flow_matrix::AbstractArray{T,2},       # the current flow matrix
+    capacity_matrix::AbstractArray{T,2}    # edge flow capacities
+    )
+    n = size(flow_graph)[1]
+
+    P = Dict{Int, Int}()                   # parent table of path
+    P[source] = -1
```

Comment:
Is there a way to do this without the (relatively slow) Dicts? Can we use forward and backwards mappings implemented via vectors?